File: bench/sieve.c
#include <stdio.h>
#include <time.h>

#define N 1000000

static unsigned char flags[N];

/* Keep the compute out-of-line so -O2 cannot fold it to a constant. */
__attribute__((noinline))
static long long sieve(void) {
    long long i, j, count;

    /* init: all prime (0) -- inside the timed region */
    for (i = 0; i < N; i++) flags[i] = 0;

    /* mark composites starting from i*i */
    for (i = 2; i * i < N; i++) {
        if (flags[i] == 0) {
            for (j = i * i; j < N; j += i) flags[j] = 1;
        }
    }

    /* count primes in [2, N) */
    count = 0;
    for (i = 2; i < N; i++) {
        if (flags[i] == 0) count++;
    }
    return count;
}

int main(void) {
    struct timespec t0, t1;
    long long result;
    long long micros;

    clock_gettime(CLOCK_MONOTONIC, &t0);
    result = sieve();
    clock_gettime(CLOCK_MONOTONIC, &t1);

    micros = (long long)(t1.tv_sec - t0.tv_sec) * 1000000LL
           + (long long)(t1.tv_nsec - t0.tv_nsec) / 1000LL;

    printf("RESULT %lld\n", result);
    printf("MICROS %lld\n", micros);
    return 0;
}

File: bench/sieve.py
import time

N = 1000000


def sieve():
    flags = bytearray(N)  # init: all prime (0) -- inside the timed region

    i = 2
    while i * i < N:
        if flags[i] == 0:
            j = i * i
            while j < N:
                flags[j] = 1
                j += i
        i += 1

    count = 0
    for i in range(2, N):
        if flags[i] == 0:
            count += 1
    return count


t0 = time.perf_counter()
result = sieve()
t1 = time.perf_counter()

micros = round((t1 - t0) * 1e6)
print("RESULT %d" % result)
print("MICROS %d" % micros)
